src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-    UnsignedInt normalIndexOffset = 1;
-    UnsignedInt textureCoordinateIndexOffset = 1;
-    _file->meshes.emplace_back(0, 0, positionIndexOffset, normalIndexOffset, textureCoordinateIndexOffset);
+void ObjImporter::parse() {
+    ArrayView<char> line = _in; /* points to beginning of current line */
+    ArrayView<char> pos = _in;  /* points to current character in line */
+
+    ObjObject* object = nullptr;
+    ObjGroup* group = nullptr;
+    ObjMeshData* meshData = nullptr;
+
+    ArrayView<char> section{nullptr};
+    int minSectionPrimitives = 0;
+    char sectionPrimitive = '?';
+
+    /* Set index 0 of data to default value */
+    _state->positions.emplace_back();
+    _state->normals.emplace_back(); // TODO Default normal?
+    _state->texCoords.emplace_back();
+
+    /* Create 'object' if not created by 'o' keyword */
+    auto ensureObject = [&]{
+        if(object == nullptr) {
+            // TODO: C++ 17
+            _state->objects.emplace_back(new ObjObject);
+            object = _state->objects.back().get();
+        }
+    };
+
+    /* Create 'group' if not created by 'g' keyword */
+    auto ensureGroup = [&]{
+        if(group == nullptr) {
+            ensureObject();
+
+            // TODO: C++ 17
+            object->groups.emplace_back(new ObjGroup{*object});
+            group = object->groups.back().get();
+        }
+    };
+
+    /* Create 'meshData' if not created by 'usemtl' keyword */
+    auto ensureMeshData = [&](){
+        if(meshData == nullptr) {
+            ensureGroup();
+            meshData = &group->meshDataForMaterial(-1);
+        }
+    };
 
-    /* The first mesh doesn't have name by default but we might find it later,
-       so we need to track whether there are any data before first name */
-    bool thisIsFirstMeshAndItHasNoData = true;
-    _file->meshNames.emplace_back();
+    /* Close a section and add it to the current meshData */
+    auto finishSection = [&]{
+        if(section.data() == nullptr) {
+            /* No open section */
+            return;
+        }
 
-    while(_file->in->good()) {
-        /* The previous object might end at the beginning of this line */
-        const std::streampos end = _file->in->tellg();
+        ensureMeshData();
+
+        ObjMesh* mesh;
+        if(sectionPrimitive == 'p') {
+            if(!meshData->points) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Points});
+                meshData->points = _state->meshes.back().get();
+            }
+            mesh = meshData->points;
+        } else if(sectionPrimitive == 'l') {
+            if(!meshData->lines) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Lines});
+                meshData->lines = _state->meshes.back().get();
+            }
+            mesh = meshData->lines;
+        } else { /* sectionPrimitive == 'f' */
```

Comment:
Just to be sure, I'd put an explicit `if(sectionPrimitive == 'f')` here and an additional `else CORRADE_ASSERT_UNREACHABLE();` at the end.